File: src/arduinobot_utils/arduinobot_utils/angle_conversion.py
#!/usr/bin/env python3
## Required os packages
# sudo apt install ros-jazzy-tf-transformations
# sudo apt install python3-transforms3d

import rclpy
from rclpy.node import Node
from arduinobot_msgs.srv import EulerToQuaternion, QuaternionToEuler
from tf_transformations import quaternion_from_euler, euler_from_quaternion

class AnglesConverter(Node):
    def __init__(self):
        super().__init__('angle_conversion_service_server')
        self.euler_to_quartenion_ = self.create_service(EulerToQuaternion, "euler_to_quaternion", self.eulerToQuarternionCallback)
        self.quartenion_to_euler_ = self.create_service(QuaternionToEuler, "quaternion_to_euler", self.quarternionToEulerCallback)
        self.get_logger().info("Angle Conversion Services Operational")

    def eulerToQuarternionCallback(self, req, res):
        self.get_logger().info(f"Converting Euler Angles - roll:{req.roll} pitch:{req.pitch} yaw:{req.yaw}")
        (res.x, res.y, res.z, res.w) = quaternion_from_euler(req.roll, req.pitch, req.yaw)
        self.get_logger().info(f"Corresponding Quaternion - x:{res.x} y:{res.y} z:{res.z} w:{res.w}")
        return res
        
    def quarternionToEulerCallback(self, req, res):
        self.get_logger().info(f"Converting Quaternion - x:{req.x} y:{req.y} z:{req.z} w:{req.w}")
        (res.roll, res.pitch, res.yaw) = euler_from_quaternion([req.x, req.y, req.z, req.w])
        self.get_logger().info(f"Corresponding Euler Angles - roll:{res.roll} pitch:{res.pitch} yaw:{res.yaw}")
        return res
    
def main():
    rclpy.init()
    angles_converter = AnglesConverter()
    rclpy.spin(angles_converter)
    angles_converter.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
    

File: src/arduinobot_utils/src/angle_conversion.cpp
#include <rclcpp/rclcpp.hpp>
#include <memory>
#include <arduinobot_msgs/srv/euler_to_quaternion.hpp>
#include <arduinobot_msgs/srv/quaternion_to_euler.hpp>
#include <tf2/utils.h>

using namespace std::placeholders;

class AnglesConversionService : public rclcpp::Node
{
public:
    AnglesConversionService() : Node("angle_conversion_service_server")
    {
        euler_to_quaternion_ = create_service<arduinobot_msgs::srv::EulerToQuaternion>("euler_to_quaternion", std::bind(&AnglesConversionService::eulerToQuaternionCallback, this, _1, _2));
        quaternion_to_euler_ = create_service<arduinobot_msgs::srv::QuaternionToEuler>("quaternion_to_euler", std::bind(&AnglesConversionService::quaternionToEulerCallback, this, _1, _2));
        RCLCPP_INFO(get_logger(), "Angle Conversion Services Operational");
    }

private:
    rclcpp::Service<arduinobot_msgs::srv::EulerToQuaternion>::SharedPtr euler_to_quaternion_;
    rclcpp::Service<arduinobot_msgs::srv::QuaternionToEuler>::SharedPtr quaternion_to_euler_;

    void eulerToQuaternionCallback(const std::shared_ptr<arduinobot_msgs::srv::EulerToQuaternion::Request> req,
                                   const std::shared_ptr<arduinobot_msgs::srv::EulerToQuaternion::Response> res)
    {
        RCLCPP_INFO_STREAM(get_logger(), "Converting Euler Angles - roll: " << req->roll << " pitch: " << req->pitch << " yaw:" << req->yaw);
        tf2::Quaternion q;
        q.setRPY(req->roll, req->pitch, req->yaw);
        res->x = q.getX();
        res->y = q.getY();
        res->z = q.getZ();
        res->w = q.getW();
        RCLCPP_INFO_STREAM(get_logger(), "Corresponding Quaternion - x:" << res->x << " y: " << res->y << " z: " << res->z << " w: " << res->w);
    }

    void quaternionToEulerCallback(const std::shared_ptr<arduinobot_msgs::srv::QuaternionToEuler::Request> req,
                                   const std::shared_ptr<arduinobot_msgs::srv::QuaternionToEuler::Response> res)
    {
        RCLCPP_INFO_STREAM(get_logger(), "Converting Quaternion - x:" << req->x << " y: " << req->y << " z: " << req->z << " w: " << req->w);
        tf2::Quaternion q(req->x,req->y,req->z,req->w);
        tf2::Matrix3x3 m(q);
        m.getRPY(res->roll,res->pitch,res->yaw);
        RCLCPP_INFO_STREAM(get_logger(), "Corresponding Euler Angles - roll: " << res->roll << " pitch: " << res->pitch << " yaw:" << res->yaw);
    }
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto anglesConversionService = std::make_shared<AnglesConversionService>();
    rclcpp::spin(anglesConversionService);
    rclcpp::shutdown();
    return 0;
}